src/Umbraco.Core/Services/TreeEntitySortingService.cs

```diff
@@ -0,0 +1,58 @@
+﻿using Umbraco.Cms.Core.Models.ContentEditing;
+using Umbraco.Cms.Core.Models.Entities;
+
+namespace Umbraco.Cms.Core.Services;
+
+public class TreeEntitySortingService : ITreeEntitySortingService
+{
+    public IEnumerable<TTreeEntity> SortEntities<TTreeEntity>(IEnumerable<TTreeEntity> entities, IEnumerable<SortingModel> sortingModels)
+        where TTreeEntity : ITreeEntity
+    {
+        TTreeEntity[] entitiesArray = entities as TTreeEntity[] ?? entities.ToArray();
+        if (entitiesArray.DistinctBy(entity => entity.ParentId).Count() != 1)
+        {
+            throw new ArgumentException("Cannot sort entities under different parents", nameof(entities));
+        }
+
+        var sortedEntityKeys = entitiesArray.OrderBy(entity => entity.SortOrder).Select(entity => entity.Key).ToList();
+
+        SortingModel[] sortingModelsArray = sortingModels as SortingModel[] ?? sortingModels.ToArray();
+
+        if (sortingModelsArray.Any(sort => sort.SortOrder < 0 || sort.SortOrder >= sortedEntityKeys.Count))
+        {
+            throw new ArgumentException("Sort order index out of range - must be within the boundaries of the entities collection", nameof(sortingModels));
```

Comment:
What happens when we delete an item?

IIRC this will create fragmentation in the sort indexes. So this will fail?

Reply:
Yep. But only in very rare circumstances (see below). That's why it's being called inside a try/catch here: https://github.com/umbraco/Umbraco-CMS/pull/14712/files#diff-c9eb3236c584a4712a8a51c833183b1510e3a394a8970169f61f74782a20e855R213

There are a fair few things that can go wrong when sorting items. Almost all of these things will be due to faulty sorting implementations on the client, which is why I have opted to only have a single error state to cover all sorting errors  (`ContentEditingOperationStatus.SortingInvalid`).

At this time there are a few rather edge-y cases where this error will end up in the client. All of these are concurrency related (multiple editors):

1. If one of the items being sorted have been deleted or moved elsewhere before the sorting operation takes place.
2. If a "sort instruction" is targeting the last position among the children of the parent, and another item among the children has been moved elsewhere or deleted. 

For outlined example in the PR description this means that the sorting operation will succeed even if some of the items have been deleted. As long as there are still 5 or more items, the sorting can still take place.